### DISCUSSION TITLE
How to bind a custom ToolPalette class

### DISCUSSION THREAD

**Discussion by rileydanejohnston** [eclipse-glsp/glsp#953] [General] [answered]
Hello,
How can I bind a CustomToolPalette after I modify/extend the default ToolPalette UIExtension? I extended the default ToolPalette so I could override the createHeader() function. How can I bind my CustomToolPalette so it gets used instead of the default ToolPalette?

**Comment by tortmayr** [comment, marked as answer]
Hi @rileydanejohnston,
the `ToolPalette` is originally bound in the [`toolPaletteModule`](https://github.com/eclipse-glsp/glsp-client/blob/v1.0.0/packages/client/src/features/tool-palette/di.config.ts):
```ts
const toolPaletteModule = new ContainerModule((bind, _unbind, isBound, rebind) => {
    bind(ToolPalette).toSelf().inSingletonScope();
    bind(TYPES.IUIExtension).toService(ToolPalette);
    configureActionHandler({ bind, isBound }, EnableToolPaletteAction.KIND, ToolPalette);
    configureActionHandler({ bind, isBound }, EnableDefaultToolsAction.KIND, ToolPalette);
});
```

If you want to use your custom implementation you simply have to `rebind` the `ToolPalette` in one of your DI modules:
```ts
const taskListDiagramModule = new ContainerModule((bind, unbind, isBound, rebind) => {
    // ...
    rebind(ToolPalette).to(MyCustomToolPalette).inSingletonScope();
});
```
